### DISCUSSION TITLE
how can i get message from outside of the ide

### DISCUSSION THREAD

**Discussion by chad-ding** [eclipse-theia/theia#14454] [Q&A]
just consider that develop a vue project via theia ide，after run "npm run serve" over terminal we can get a preview url，but how can i get the url inside ide then display it. currently i get from terminal log with regexp pattern, but it's absoule such a bad idea.

**Comment by msujew** [comment]
Hey @chad-ding,

you need some sort of port-forwarding service. Something that continuously scans the system to see whether any new ports have opened and that creates a new reverse proxy on the service layer (the last part is optional, as it depends on how you deploy your web IDE). Ideally this service would communicate with your IDE to display all available ports for easy access. Note that this is not trivial. I believe that [Gitpod](https://github.com/gitpod-io/gitpod) has an open source implementation of such a service somewhere.

**Reply by chad-ding** [reply]
thanks for answer, it's an option but still not so good, it means ide needs to scan ports over an over, otherwise ide may miss the state whe web server stopped, electron seems can communicate with schema protocol, not sure it's usefull for theia or not

**Reply by msujew** [reply]
> it's an option but still not so good, it means ide needs to scan ports over an over, otherwise ide may miss the state whe web server stopped

Yes, it's not ideal, but as long as OS vendors don't give developers a native way of hooking into network events (like the opening of ports), it's the way to go. Polling information like this every few seconds shouldn't put any strain on the system.

> electron seems can communicate with schema protocol, not sure it's usefull for theia or not

I'm not sure how this is relevant?

**Reply by chad-ding** [reply]
ok,  i will try to start a http server inside ide, it may works
